src/core/thread/link_metrics.cpp

```diff
@@ -44,39 +44,177 @@
+    aLinkMetrics.mLqi        = mLinkMetrics.mLqi;
+    aLinkMetrics.mLinkMargin = mLinkMetrics.mLinkMargin;
+    aLinkMetrics.mRssi       = mLinkMetrics.mRssi;
+}
+
+bool LinkMetricsSeriesInfo::IsFrameTypeMatch(uint8_t aFrameType) const
+{
+    bool match = false;
+
+    switch (aFrameType)
+    {
+    case kSeriesTypeLinkProbe:
+        VerifyOrExit(!mSeriesFlags.IsMacDataFlagSet()); // Ignore this when Mac Data is accounted
+        match = mSeriesFlags.IsLinkProbeFlagSet();
+        break;
+    case Mac::Frame::kFcfFrameData:
+        match = mSeriesFlags.IsMacDataFlagSet();
+        break;
+    case Mac::Frame::kFcfFrameMacCmd:
+        match = mSeriesFlags.IsMacDataRequestFlagSet();
+        break;
+    case Mac::Frame::kFcfFrameAck:
+        match = mSeriesFlags.IsMacAckFlagSet();
+        break;
+    default:
+        break;
+    }
+
+exit:
+    return match;
+}
+
 LinkMetrics::LinkMetrics(Instance &aInstance)
     : InstanceLocator(aInstance)
     , mLinkMetricsReportCallback(nullptr)
     , mLinkMetricsReportCallbackContext(nullptr)
+    , mLinkMetricsSeriesInfoPool()
 {
 }
 
 otError LinkMetrics::LinkMetricsQuery(const Ip6::Address & aDestination,
                                       uint8_t              aSeriesId,
-                                      const otLinkMetrics &aLinkMetricsFlags)
+                                      const otLinkMetrics *aLinkMetricsFlags)
 {
     otError                error;
     LinkMetricsTypeIdFlags typeIdFlags[kMaxTypeIdFlags];
-    uint8_t                typeIdFlagsCount = TypeIdFlagsFromLinkMetricsFlags(typeIdFlags, aLinkMetricsFlags);
+    uint8_t                typeIdFlagsCount = 0;
+
+    if (aLinkMetricsFlags != nullptr)
+    {
+        typeIdFlagsCount = TypeIdFlagsFromLinkMetricsFlags(typeIdFlags, *aLinkMetricsFlags);
+    }
+
+    if (aSeriesId != 0)
+    {
+        VerifyOrExit(typeIdFlagsCount == 0, error = OT_ERROR_INVALID_ARGS);
+    }
 
     error = SendLinkMetricsQuery(aDestination, aSeriesId, typeIdFlags, typeIdFlagsCount);
 
+exit:
     return error;
 }
 
-otError LinkMetrics::AppendLinkMetricsReport(Message &aMessage, const Message &aRequestMessage)
+otError LinkMetrics::SendMgmtRequestForwardTrackingSeries(const Ip6::Address &            aDestination,
+                                                          uint8_t                         aSeriesId,
+                                                          const otLinkMetricsSeriesFlags &aSeriesFlags,
+                                                          const otLinkMetrics &           aLinkMetricsFlags)
 {
-    otError       error = OT_ERROR_NONE;
-    Tlv           tlv;
-    uint8_t       queryId;
-    bool          hasQueryId  = false;
-    uint8_t       length      = 0;
-    uint16_t      startOffset = aMessage.GetLength();
-    uint16_t      offset;
-    uint16_t      endOffset;
-    otLinkMetrics linkMetrics;
+    otError     error = OT_ERROR_NONE;
+    Tlv         forwardProbingRegistrationSubTlv;
```

Comment:
is it necessary? may use a pointer of type TLV to the start of `subTLVs` buffer to avoid below unnecessary memcpy below?

Reply:
I think some compiler would complain about such practice: https://github.com/openthread/openthread/issues/5647. Besides, I think code like this is not quite readable.

Reply:
Have updated.